### PR TITLE
Render ArgoCD using kustomize and install konfigure plugin

### PR DESCRIPTION
**Pull request by kubasobon**
Fixes https://github.com/giantswarm/giantswarm/issues/16947
Fixes https://github.com/giantswarm/giantswarm/issues/17169
Fixes https://github.com/giantswarm/giantswarm/issues/17220
Replaces https://github.com/giantswarm/management-clusters-fleet/pull/6

This is a big PR. I couldn't find a good way to slice it and still present all
the changes in comprehensible format. LMK if anything seems off/not as we
discussed.

### CODE REVIEW

**Comment by corest** [comment]
Did you test this on any test installation?

**Comment by kubasobon** [comment]
Not yet. Which one would be good for that?

**Comment by corest** [comment]
Any you prefer. Just check if argo is installed there already to not mess with existing installation

File: hack/update-argocd-manifests.sh [hunk 1]
```diff
@@ -12,26 +12,16 @@ KUSTOMIZE=kustomize
...
-cd -
-
-sed -i "/dex/d" manifests/base/kustomization.yaml
-
-mkdir -p ${SRCROOT}/build/argocd
-echo "${AUTOGENMSG}" > "${SRCROOT}/build/argocd/install.yaml"
-${KUSTOMIZE} build "manifests/cluster-install" >> "${SRCROOT}/build/argocd/install.yaml"
+mkdir -p ${SRCROOT}/manifests/bases/upstream-argocd
+echo "${AUTOGENMSG}" > "${SRCROOT}/manifests/bases/upstream-argocd/install.yaml"
+${KUSTOMIZE} build "manifests/cluster-install" >> "${SRCROOT}/manifests/bases/upstream-argocd/install.yaml"
```
Comment (kopiczko) [inline hack/update-argocd-manifests.sh:24]: This script is melting :D But that's good 👍

File: manifests/argocd/deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app.kubernetes.io/component: repo-server
+    app.kubernetes.io/name: argocd-repo-server
+    app.kubernetes.io/part-of: argocd
```
Comment (kopiczko) [inline manifests/argocd/deploy.yaml:7]: Again the labels should be removed.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+manifests/bases/upstream-argocd
+!manifests/bases/upstream-argocd/kustomization.yaml
```
Comment (kopiczko) [inline .gitignore:2]: I think we should commit `install.yaml` as well for reproducibility.

Comment (kopiczko) [inline .gitignore:2]: BTW I can't find manifests/bases/upstream-argocd/kustomization.yaml in this PR.

File: manifests/provider/vmware/kustomization.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-resources:
-- ../../base
+bases:
+- ../../bases/argocd
```
Comment (kopiczko) [inline manifests/provider/vmware/kustomization.yaml:2]: Shouldn't that be `../../argocd`?

File: .gitignore (hunk 3 above)
Comment (kubasobon) [reply .gitignore]: Missed this, fixed.

File: manifests/provider/vmware/kustomization.yaml (hunk 4 above)
Comment (kubasobon) [reply manifests/provider/vmware/kustomization.yaml:2]: I think so. This would simplify the script as well.

Comment (kopiczko) [reply manifests/provider/vmware/kustomization.yaml:2]: But `../../bases/argocd` doesn't even exist. Am I wrong?

It's either `../../argocd` or `../../bases/argocd-upstream`

Comment (kubasobon) [reply manifests/provider/vmware/kustomization.yaml]: It was supposed to exist, but I got rid of it in the process. It's `../../argocd`.

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+manifests/bases/upstream-argocd/*
+!manifests/bases/upstream-argocd/kustomization.yaml
+!manifests/bases/upstream-argocd/install.yaml
```
Comment (kopiczko) [inline .gitignore:3]: What is then ignored here?

File: Makefile.custom.mk [hunk 6]
```diff
@@ -12,6 +12,8 @@ build/argocd: ARGOCD_VERSION=v2.0.1
 build/argocd:
 	@echo "====> $@"
 	./hack/update-argocd-manifests.sh $(ARGOCD_VERSION)
+	echo "$(AUTOGENMSG)" > build/argocd/install.yaml
+	$(KUSTOMIZE) build manifests/argocd >> build/argocd/install.yaml
```
Comment (kopiczko) [inline Makefile.custom.mk:16]: I wonder if we need to build `build/argocd` at all. All that's there should be included in `build/provider/*` right?

Comment (kubasobon) [reply Makefile.custom.mk:16]: Yeah. Maybe we don't need two steps then, but just produce the final `build/provider/*` files?

Comment (kopiczko) [reply Makefile.custom.mk:16]: IMO that would be ideal.

Comment (kopiczko) [reply Makefile.custom.mk:16]: I think we also can skip the `/provider/` part then.

File: manifests/argocd/deploy.yaml (hunk 2 above)
Comment (kopiczko) [reply manifests/argocd/deploy.yaml:7]: They are still there.

File: manifests/argocd/argocd-repo-server-deploy.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      containers:
+      - name: argocd-repo-server
+        volumeMounts:
+        - mountPath: /usr/local/bin/konfigure
+          name: giantswarm-bin
+          subPath: konfigure
+      initContainers:
+      - args:
+        - --vault-address=$(VAULT_ADDR)
+        - --vault-role=konfigure
```
Comment (kubasobon) [inline manifests/argocd/argocd-repo-server-deploy.yaml:22]: @corest relevant to role/policy rename ticket

**Review by corest — APPROVED** [review]
LGTM!

**Comment by kopiczko** [comment]
This got outdated https://github.com/giantswarm/management-clusters-fleet#layout

**Comment by kubasobon** [comment]
@kopiczko I'll push README.md update in a sec.

File: README.md [hunk 8]
```diff
@@ -10,7 +10,6 @@ GitOps for Giant Swarm management clusters
     - `collections` AppProject
   - Application: `argocd` manages:
     - Upstream Argo CD manifests stored in `build/argocd/`
-- AppProject: `collections`:
```
Comment (corest) [inline README.md:13]: We still have AppProject `collections`. We don't have app `projects` now. So layout now looks like this:

```
- AppProject: `argocd` contains:
  - Application: `argocd` manages:
    -  upstream Argo CD manifests
    - `argocd` AppProject CR
    - `argocd` Application CR
    - `collections` AppProject CR
    - `provider-collection` Application CR
  - Application: `provider-collection` manages:
    - Manifests from the provider-specific app collection
```

**Comment by kubasobon** [comment]
Updated layout with changes proposed by @corest. Merging.

**Pull request merged**